storage/remote/queue_manager.go

```diff
@@ -1028,7 +1032,90 @@ func (s *shards) enqueue(ref chunks.HeadSeriesRef, data interface{}) bool {
 	}
 }
 
-func (s *shards) runShard(ctx context.Context, shardID int, queue chan interface{}) {
+type queue struct {
+	batch      []interface{}
+	batchQueue chan []interface{}
+
+	// Since we know there is a limited number of batches out, using a stack is
+	// easy and persistent across GC compared to a sync.Pool
+	batchPool [][]interface{}
+	poolMux   sync.Mutex
+}
+
+func newQueue(batchSize, capacity int) *queue {
+	batches := capacity / batchSize
+	return &queue{
+		batch:      make([]interface{}, 0, batchSize),
+		batchQueue: make(chan []interface{}, batches),
+		batchPool:  make([][]interface{}, 0, batches),
+	}
+}
+
+func (q *queue) Append(datum interface{}, stop <-chan struct{}) bool {
+	q.batch = append(q.batch, datum)
+	if len(q.batch) == cap(q.batch) {
```

Comment:
Data waits on this side of the chan until we have a full batch?
That is different semantics to before, and breaks the contract of `BatchSendDeadline`.

Reply:
It will wait on this side of the channel with the exception of if `BatchSendDeadline` is hit. When the deadline is hit I take the lock on the `shards` struct to block new appends and get the partial batch directly. This felt a bit hacky to me, but I believe it keeps the semantics the same and avoided needing a lock on every sample.

I did just realize there was a potential race condition where if a batch filled while acquiring the lock on `shards` it could send out of order timestamps and fixed that. Now when `BatchSendDeadline` is hit it will try to read from `batchQueue` and only get the current batch if there is nothing in `batchQueue`.

Reply:
Thanks, I missed that. Also that the batch is flushed on a reshard.